Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp

```diff
@@ -0,0 +1,349 @@
+    if (!localizationJSON)
+        return emptyString();
+
+    RefPtr stringJSON = localizationJSON->getObject(key.convertToASCIILowercase());
+    if (!stringJSON)
+        return emptyString();
+
+    auto localizedString = stringJSON->getString(messageKey);
+    if (!localizedString.length())
+        return emptyString();
+
+    RefPtr namedPlaceholders = JSONWithLowercaseKeys(stringJSON->getObject(placeholdersKey));
+
+    localizedString = stringByReplacingNamedPlaceholdersInString(localizedString, namedPlaceholders);
+    localizedString = stringByReplacingPositionalPlaceholdersInString(localizedString, placeholders);
+
+    localizedString = localizedString.impl()->replace("$$"_s, "$"_s);
+
+    return localizedString;
+}
+
+void WebExtensionLocalization::localizedArrayForArray(RefPtr<JSON::Array> json)
+{
+    if (!json)
+        return;
+
+    for (Ref value : *json) {
+        auto valueType = value->type();
+
+        if (valueType == JSON::Value::Type::String)
+            value->asString() = localizedStringForString(value->asString());
+        else if (valueType == JSON::Value::Type::Array)
+            localizedArrayForArray(value->asArray());
+        else if (valueType == JSON::Value::Type::Object)
+            localizedJSONforJSON(value->asObject());
+    }
+}
+
+String WebExtensionLocalization::localizedStringForString(String sourceString)
+{
+    String localizedString = sourceString;
+
+    auto localizableStringRegularExpression = JSC::Yarr::RegularExpression("__MSG_([A-Za-z0-9_@]+)__"_s, { });
+
+    int index = 0;
+    while (index < static_cast<int>(localizedString.length())) {
+        int matchLength;
+        index = localizableStringRegularExpression.match(localizedString, index, &matchLength);
+        if (index < 0)
+            break;
+
+        auto key = localizedString.substring(index, matchLength);
+        auto localizedReplacement = localizedStringForKey(key);
+
+        localizedString = makeStringByReplacingAll(localizedString, key, localizedReplacement);
+
+        index += localizedReplacement.length();
+    }
+
+    return localizedString;
+}
+
+RefPtr<JSON::Object> WebExtensionLocalization::localizationJSONForWebExtension(WebExtension& webExtension, const String& withLocale)
+{
+    StringBuilder pathBuilder;
+    pathBuilder.append("_locales/"_s);
+    pathBuilder.append(withLocale);
+    pathBuilder.append("/messages.json"_s);
+
+    auto path = pathBuilder.toString();
+
+    RefPtr<API::Error> error;
+    RefPtr data = webExtension.resourceDataForPath(path, error, WebExtension::CacheResult::No, WebExtension::SuppressNotFoundErrors::Yes);
+    if (!data || error) {
+        webExtension.recordErrorIfNeeded(error);
+        return nullptr;
+    }
+
+    auto json = JSON::Value::parseJSON(String::fromUTF8(data->span()));
+
+    return json ? json->asObject() : nullptr;
+}
+
+Ref<JSON::Object> WebExtensionLocalization::predefinedMessages(RefPtr<JSON::Object> localizedJSON)
```

Comment:
Param not needed anymore.
```suggestion
Ref<JSON::Object> WebExtensionLocalization::predefinedMessages()
```